Automattic/wp-calypso
# Plans: Convert getPlansBySite to TypeScript

#### Proposed Changes

This PR converts the `getPlansBySite()` and `getPlansBySiteId()` Redux selectors to TypeScript and fixes the one existing type that was already describing their results to clarify that it can include null data. This should not have any functional effect.

To verify that the new types are accurate, this also converts the `StarterPlanDetails` component to TypeScript which uses the data from the modified functions ([via a selector in CheckoutThankYou](https://github.com/Automattic/wp-calypso/blob/1e2bb63052b40d4191c2e682286d241e8605838a/client/my-sites/checkout/checkout-thank-you/index.jsx#L736)).

#### Screenshots

The `StarterPlanDetails` component:

<img width="839" alt="Screen Shot 2022-09-28 at 12 37 37 PM" src="https://user-images.githubusercontent.com/2036909/192838882-cabace5f-45d4-495c-899e-e3eeec68af86.png">

The `CustomDomainPurchaseDetail` component which is actually the consumer of the data from `getPlansBySite()`.

<img width="831" alt="Screen Shot 2022-09-28 at 12 42 41 PM" src="https://user-images.githubusercontent.com/2036909/192841169-7c1a2088-c56e-45fb-9308-1aa2f686398d.png">

#### Testing Instructions

- On a site with no active plan (and therefore an available domain credit), visit `/checkout/starter-plan`.
- Complete the purchase.
- Verify that the post-checkout page renders successfully.
- Using React devtools, verify that the `CustomDomainPurchaseDetail` component has a `hasDomainCredit` prop set to `true`. This proves that the `plan` object was successfully fetched using `getPlansBySite()`.

<img width="198" alt="Screen Shot 2022-09-28 at 12 42 54 PM" src="https://user-images.githubusercontent.com/2036909/192841584-ef545d12-e97b-4262-a773-77d718d8200c.png">
